readme.md
# Postman Monitor with Prometheus

This project provides a small Node.js based server which will run a Postman collection continuously on regular interval. It uses the Newman library that Postman provides, and exposes various HTTP and request metrics and stats in Prometheus exposition format so they can be scraped by Prometheus. This allows you to proactively monitor any Postman collection you wish, from wherever you wish.

You can use this to monitor the health and performance of web sites and REST APIs, and by using test assertions (see below) you can validate the contents returned, status codes, headers etc. Even chain requests together extracting variables from one request to use with another.

By default the server listens on port 8080 and provides metrics at the standard `/metrics` endpoint. The collection you want to run can be fetched by the runner at startup from a URL you supply, or you can build the runner container with the collection file copied into the image.

The server is containerised and published publicly at ghcr.io/benc-uk/postman-prometheus

Kubernetes manifests for deployment are also provided.

Goals:

- Turn Postman into a continuous monitoring tool
- Export Postman/Newman data into Prometheus

Use cases & key features:

- Monitoring of APIs and websites
- Go beyond simple single HTTP checks, using Postman's [collection feature](https://learning.postman.com/docs/running-collections/intro-to-collection-runs/) and [test assertions](https://learning.postman.com/docs/writing-scripts/script-references/test-examples/) for checking and chaining

![](https://img.shields.io/github/license/benc-uk/postman-prometheus)
![](https://img.shields.io/github/last-commit/benc-uk/postman-prometheus)
![](https://img.shields.io/github/release/benc-uk/postman-prometheus)
![](https://img.shields.io/github/checks-status/benc-uk/postman-prometheus/main)

# Example

This is a screenshot of the provided Grafana dashboard

The dashboard can be found in the [grafana directory](https://github.com/benc-uk/postman-prometheus/tree/main/grafana) as a JSON file which you can import.

![screen shot of dashboard](https://user-images.githubusercontent.com/14982936/111913204-fd8a6400-8a64-11eb-95c1-a40f4828d05f.png)

# Config

| Environmental Variable | Purpose                                                          | Default           |
| ---------------------- | ---------------------------------------------------------------- | ----------------- |
| PORT                   | Port the server listens on                                       | 8080              |
| COLLECTION_FILE        | Path to Postman collection file to load and use                  | ./collection.json |
| COLLECTION_URL         | Load from Postman collection from URL, overrides COLLECTION_FILE | _none_            |
| RUN_INTERVAL           | How frequently to run the collection, in seconds                 | 30                |
| RUN_ITERATIONS         | How many iterations of the collection to run                     | 1                 |
| ENABLE_BAIL            | Stops the run when a test case or request fails                  | false             |
| ENABLE_REQUEST_METRICS | Disable the per-request metrics if you wish                      | true              |
| ENVIRONMENT_FILE       | Path to a Postman environment file                               | _none_            |
| ENVIRONMENT_URL        | Load a Postman environment from URL, overrides ENVIRONMENT_FILE  | _none_            |
| POSTMAN\_{varname}     | Environment vars to pass to running the collection               | _none_            |

## Note on Postman variables

Postman/Newman can [accept variables a number of ways](https://learning.postman.com/docs/sending-requests/variables/) with this runner you supply values for any variables your scripts reference in two ways:

- Environments file, created by defining an environment in Postman and [exporting as JSON](https://learning.postman.com/docs/getting-started/importing-and-exporting-data/#exporting-environments)
- Using special `POSTMAN_{varname}` environment vars, set as regular OS environment vars (therefor settable at runtime from Docker and Kubernetes). The prefix `POSTMAN_` is required and stripped off, leaving the name of the variable to set when running the collection, e.g. if your Postman request referenced a variable `{{myvar}}` you can set it using `POSTMAN_myvar=foo`

# Repo Contents

📁 `src` - Source of the Node.js runner which wraps Newman  
📁 `samples` - Example Postman collections for monitoring  
📁 `grafana` - Example Grafana dashboard which can be imported  
📁 `deploy` - Deployment to Kubernetes, plus Helm samples for alerting  
📁 `build` - Dockerfile mostly  
📁 `scripts` - Some bash scripts

# Using

Just about everything you need to do with this project has been put into make

```txt
$ make

help                 💬 This help message 
run                  🥈 Run locally (requires Node.js) ‍
image                🔨 Build container image from Dockerfile 
lint-fix             📜 Lint & format, will try to fix errors and modify code 
lint                 🔎 Lint & format, will not fix but sets exit code on error 
push                 📤 Push container image to registry
deploy               🚀 Deploy to Kubernetes 
undeploy             💀 Remove from Kubernetes 
```

The `deploy` target provides a lightweight "Helm-less" way to deploy to Kubernetes, using envsubst, makefile variables and deploy/deployment.yaml as a template.
Before running `make deploy` check the `DEPLOY_` variables in the makefile, then either edit or override these

# Exported Metrics

Below is a dump of all the metrics the server will export.

> **💬 Note**. There could be MANY per-request metrics, if running a collection with a large number of requests and a large number of iterations. You have been warned!

```txt
# ==========================================================
# Lifetime metrics for the whole life of the monitor server
# ==========================================================

# TYPE postman_lifetime_runs_total counter
postman_lifetime_runs_total{collection="Example"} 69

# TYPE postman_lifetime_iterations_total counter
postman_lifetime_iterations_total{collection="Example"} 138

# TYPE postman_lifetime_requests_total counter
postman_lifetime_requests_total{collection="Example"} 276

# ==========================================================
# Metrics aggregated for the whole collection
# ==========================================================

# TYPE postman_stats_iterations_total gauge
postman_stats_iterations_total{collection="Example"} 2

# TYPE postman_stats_iterations_failed gauge
postman_stats_iterations_failed{collection="Example"} 0 

# TYPE postman_stats_requests_total gauge
postman_stats_requests_total{collection="Example"} 4

# TYPE postman_stats_requests_failed gauge
postman_stats_requests_failed{collection="Example"} 0

# TYPE postman_stats_tests_total gauge
postman_stats_tests_total{collection="Example"} 4

# TYPE postman_stats_tests_failed gauge
postman_stats_tests_failed{collection="Example"} 0

# TYPE postman_stats_test_scripts_total gauge
postman_stats_test_scripts_total{collection="Example"} 4

# TYPE postman_stats_test_scripts_failed gauge
postman_stats_test_scripts_failed{collection="Example"} 0

# TYPE postman_stats_assertions_total gauge
postman_stats_assertions_total{collection="Example"} 8

# TYPE postman_stats_assertions_failed gauge
postman_stats_assertions_failed{collection="Example"} 2

# TYPE postman_stats_transfered_bytes_total gauge
postman_stats_transfered_bytes_total{collection="Example"} 43750

# TYPE postman_stats_resp_avg gauge
postman_stats_resp_avg{collection="Example"} 40.25

# TYPE postman_stats_resp_min gauge
postman_stats_resp_min{collection="Example"} 17

# TYPE postman_stats_resp_max gauge
postman_stats_resp_max{collection="Example"} 107

# ==========================================================
# These metrics are per request AND per iteration
# ==========================================================

# TYPE postman_request_status_code gauge
postman_request_status_code{request_name="Some Example Request",iteration="0",collection="Example"} 200

# TYPE postman_request_resp_time gauge
postman_request_resp_time{request_name="Some Example Request",iteration="0",collection="Example"} 19

# TYPE postman_request_resp_size gauge
postman_request_resp_size{request_name="Some Example Request",iteration="0",collection="Example"} 8684

# TYPE postman_request_status_ok gauge
postman_request_status_ok{request_name="Some Example Request",iteration="0",collection="Example"} 1

# TYPE postman_request_failed_assertions gauge
postman_request_failed_assertions{request_name="Some Example Request",iteration="0",collection="Example"} 0

# TYPE postman_request_total_assertions gauge
postman_request_total_assertions{request_name="Some Example Request",iteration="0",collection="Example"} 3
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.2.1] - 2021-11-09

- Logging of runs and assertion failures to stdout 

## [1.2.0] - 2021-11-04

- Now emits per-request assertion metrics (total and failed)

## [1.1.0] - 2021-03-30

- Support for setting vars at runtime using `POSTMAN_{var}`

## [1.0.0] - 2021-03-19

- First release



src/http.js
/*
  Simple compact HTTP library in vanilla Node.js
  Ben Coleman, 2020
*/

const URL = require('url')

module.exports = class HTTP {
  constructor(base, parseResults = true, auth = null, headers = {}, checkStatus = true) {
    this.baseUrl = base
    this.parseResults = parseResults
    this.checkStatus = checkStatus

    if (auth) {
      if (!auth.creds) {
        throw 'HTTP error: auth creds must be set to `token` or `user:password`'
      }
      switch (auth.type.toLowerCase()) {
        case 'bearer':
          this.baseHeaders = { Authorization: `Bearer ${auth.creds}`, ...headers }
          break
        case 'basic': {
          let basicAuthBuff = Buffer.from(auth.creds)
          this.baseHeaders = { Authorization: `Basic ${basicAuthBuff.toString('base64')}`, ...headers }
          break
        }
        default:
          throw "HTTP error: auth type must be 'basic' or 'bearer'"
      }
    } else {
      this.baseHeaders = headers
    }
  }

  // HTTP GET wrapper
  async get(path, headers = {}) {
    return await this.request(path, 'GET', headers)
  }

  // HTTP POST wrapper
  async post(path, data, contentType = 'application/json', headers = {}) {
    const reqBody = typeof data == 'object' ? JSON.stringify(data) : data
    return await this.request(path, 'POST', { 'Content-Type': contentType, ...headers }, reqBody)
  }

  // HTTP PUT wrapper
  async put(path, data, contentType = 'application/json', headers = {}) {
    const reqBody = typeof data == 'object' ? JSON.stringify(data) : data
    return await this.request(path, 'PUT', { 'Content-Type': contentType, ...headers }, reqBody)
  }

  // HTTP DELETE wrapper
  async delete(path, data, contentType = 'application/json', headers = {}) {
    const reqBody = typeof data == 'object' ? JSON.stringify(data) : data
    return await this.request(path, 'DELETE', { 'Content-Type': contentType, ...headers }, reqBody)
  }

  // Generic low level HTTP request wrapped as a promise
  request(path, method = 'GET', headers = {}, reqBody = null) {
    let req = URL.parse(`${this.baseUrl}${path}`)
    req.headers = { ...this.baseHeaders, ...headers }
    req.method = method

    return new Promise((resolve, reject) => {
      const httpLib = req.protocol && req.protocol.startsWith('https') ? require('https') : require('http')

      if (reqBody) {
        req.headers['Content-Length'] = reqBody.length
      }

      if (this.debug) {
        console.log('### HTTP client request:', req, '\n### HTTP client body:', reqBody)
      }

      const request = httpLib.request(req, (response) => {
        let body = []
        response.on('data', (chunk) => body.push(chunk))
        response.on('end', () => {
          let data
          try {
            data = this.parseResults ? JSON.parse(body.join('')) : body.join('')
          } catch (err) {
            reject(`JSON parsing of response body failed  - ${err}`)
          }

          // Our custom response result wrapper, with Axios like fields
          const resp = {
            headers: response.headers,
            status: response.statusCode,
            statusText: response.statusMessage,
            data,
          }

          if (this.debug) {
            console.log(`### HTTP client response:`, resp)
          }
          if (this.checkStatus && (response.statusCode < 200 || response.statusCode > 299)) {
            console.error(`Error! Request failed with status code: ${response.statusCode}`)
            reject(resp)
          }
          resolve(resp)
        })
      })

      request.on('error', (err) => reject(err))
      if (reqBody) {
        request.write(reqBody)
      }
      request.end()
    })
  }
}


scripts/test-alert-manager.sh
#!/bin/bash

# Taken from https://gist.github.com/cherti/61ec48deaaab7d288c9fcf17e700853a

name=$RANDOM
url='http://localhost:9093/api/v1/alerts'

echo "### Firing up alert $name" 

curl -XPOST $url -d "[{ 
	\"status\": \"firing\",
	\"labels\": {
		\"alertname\": \"$name\",
		\"service\": \"my-service\",
		\"severity\":\"warning\",
		\"instance\": \"$name.example.net\"
	},
	\"annotations\": {
		\"summary\": \"High latency is high!\"
	},
	\"generatorURL\": \"http://prometheus.int.example.net/<generating_expression>\"
}]"

echo ""

echo "### Press enter to resolve alert"
read

echo "### Sending resolve"
curl -XPOST $url -d "[{ 
	\"status\": \"resolved\",
	\"labels\": {
		\"alertname\": \"$name\",
		\"service\": \"my-service\",
		\"severity\":\"warning\",
		\"instance\": \"$name.example.net\"
	},
	\"annotations\": {
		\"summary\": \"High latency is high!\"
	},
	\"generatorURL\": \"http://prometheus.int.example.net/<generating_expression>\"
}]"

echo ""

src/server.js
const newman = require('newman')
const express = require('express')
const fs = require('fs')
const http = require('./http.js')

let collectionFile = process.env.COLLECTION_FILE || './collection.json'
const collectionUrl = process.env.COLLECTION_URL || ''
const envFile = process.env.ENVIRONMENT_FILE || ''
const envUrl = process.env.ENV_URL || ''
const port = process.env.PORT || '8080'
const runInterval = process.env.RUN_INTERVAL || '30'
const runIterations = process.env.RUN_ITERATIONS || '1'
const enableBail = process.env.ENABLE_BAIL || 'false'
const requestMetrics = process.env.ENABLE_REQUEST_METRICS || 'true'

let collectionName = ''
let resultSummary = {}

// Lifetime global counters
let runCount = 0
let iterationCount = 0
let reqCount = 0

//
// Entrypoint and server startup is here....
//

const app = express()

app.get('/metrics', (req, res) => {
  res.setHeader('content-type', 'text/plain; charset=utf-8; version=0.0.4')

  let metricString = ''
  try {
    metricString = addMetric(metricString, 'lifetime_runs_total', runCount, 'counter')
    metricString = addMetric(metricString, 'lifetime_iterations_total', iterationCount, 'counter')
    metricString = addMetric(metricString, 'lifetime_requests_total', reqCount, 'counter')
    metricString = addMetric(metricString, 'stats_iterations_total', resultSummary.run.stats.iterations.total)
    metricString = addMetric(metricString, 'stats_iterations_failed', resultSummary.run.stats.iterations.failed)
    metricString = addMetric(metricString, 'stats_requests_total', resultSummary.run.stats.requests.total)
    metricString = addMetric(metricString, 'stats_requests_failed', resultSummary.run.stats.requests.failed)
    metricString = addMetric(metricString, 'stats_tests_total', resultSummary.run.stats.tests.total)
    metricString = addMetric(metricString, 'stats_tests_failed', resultSummary.run.stats.tests.failed)
    metricString = addMetric(metricString, 'stats_test_scripts_total', resultSummary.run.stats.testScripts.total)
    metricString = addMetric(metricString, 'stats_test_scripts_failed', resultSummary.run.stats.testScripts.failed)
    metricString = addMetric(metricString, 'stats_assertions_total', resultSummary.run.stats.assertions.total)
    metricString = addMetric(metricString, 'stats_assertions_failed', resultSummary.run.stats.assertions.failed)
    metricString = addMetric(metricString, 'stats_transfered_bytes_total', resultSummary.run.transfers.responseTotal)
    metricString = addMetric(metricString, 'stats_resp_avg', resultSummary.run.timings.responseAverage)
    metricString = addMetric(metricString, 'stats_resp_min', resultSummary.run.timings.responseMin)
    metricString = addMetric(metricString, 'stats_resp_max', resultSummary.run.timings.responseMax)

    if (requestMetrics == 'true') {
      for (let execution of resultSummary.run.executions) {
        if (!execution.response) {
          continue
        }

        const labels = [
          {
            // eslint-disable-next-line camelcase
            request_name: execution.item.name,
          },
          {
            iteration: execution.cursor.iteration,
          },
        ]
        if (execution.response.code) {
          metricString = addMetric(metricString, 'request_status_code', execution.response.code, 'gauge', labels)
        }
        if (execution.response.responseTime) {
          metricString = addMetric(metricString, 'request_resp_time', execution.response.responseTime, 'gauge', labels)
        }
        if (execution.response.responseSize) {
          metricString = addMetric(metricString, 'request_resp_size', execution.response.responseSize, 'gauge', labels)
        }
        if (execution.response.status) {
          const statusOK = execution.response.status == 'OK' ? 1 : 0
          metricString = addMetric(metricString, 'request_status_ok', statusOK, 'gauge', labels)
        }

        let failedAssertions = 0
        let totalAssertions = 0
        // Include per request assertion metrics
        if (execution.assertions) {
          for (let a in execution.assertions) {
            totalAssertions++
            if (execution.assertions[a].error) {
              failedAssertions++
            }
          }
        }
        metricString = addMetric(metricString, 'request_failed_assertions', failedAssertions, 'gauge', labels)
        metricString = addMetric(metricString, 'request_total_assertions', totalAssertions, 'gauge', labels)
      }
    }

    res.send(metricString)
  } catch (err) {
    res.status(500).send('No result data to show, maybe the collection has not run yet')
  }
})

app.get('/', (req, res) => {
  res.setHeader('content-type', 'text/plain')
  res.status(404).send('Nothing here, try /metrics')
})

app.listen(port, async () => {
  // COLLECTION_URL when set takes priority over COLLECTION_FILE
  if (collectionUrl) {
    logMessage(`Collection URL will be fetched and used ${collectionUrl}`)
    try {
      const httpClient = new http(collectionUrl, false)
      let resp = await httpClient.get('')
      fs.writeFileSync(`./downloaded-collection.tmp.json`, resp.data)
      // Note. Overwrite the COLLECTION_FILE setting if it was already set
      collectionFile = './downloaded-collection.tmp.json'
    } catch (err) {
      logMessage(`FATAL! Failed to download collection from URL\n ${JSON.stringify(err, null, 2)}`)
      process.exit(1)
    }
  }

  // ENV_URL when set takes priority over ENVIRONMENT_FILE
  if (envUrl) {
    logMessage(`Postman Environment file URL will be fetched and used ${envUrl}`)
    try {
      const httpClient = new http(envUrl, false)
      let resp = await httpClient.get('')
      fs.writeFileSync(`./downloaded-env.tmp.json`, resp.data)
      // Note. Overwrite the ENVIRONMENT_FILE setting if it was already set
      envFile = './downloaded-env.tmp.json'
    } catch (err) {
      logMessage(`FATAL! Failed to download environment from URL\n ${JSON.stringify(err, null, 2)}`)
      process.exit(1)
    }
  }

  if (!fs.existsSync(collectionFile)) {
    logMessage(`FATAL! Collection file '${collectionFile}' not found`)
    process.exit(1)
  }

  logMessage(`Newman runner started & listening on ${port}`)
  logMessage(`Collection will be run every ${runInterval} seconds`)

  runCollection()
  setInterval(runCollection, parseInt(runInterval * 1000))
})

//
// Monitoring and Prometheus functions
//

function runCollection() {
  logMessage(`Starting run of ${collectionFile}`)

  // Special logic to bring all env vars starting with POSTMAN_ into the run
  let postmanEnvVar = []
  for (let envVar in process.env) {
    if (envVar.startsWith('POSTMAN_')) {
      postmanEnvVar.push({
        // Remove the prefix
        key: envVar.replace('POSTMAN_', ''),
        value: process.env[envVar],
      })
    }
  }

  newman.run(
    {
      collection: require(collectionFile),
      iterationCount: parseInt(runIterations),
      bail: enableBail == 'true',
      environment: envFile,
      envVar: postmanEnvVar,
    },
    runComplete
  )
}

function runComplete(err, summary) {
  if (!summary) {
    logMessage(`ERROR! Failed to run collection, no summary was returned!`)
    return
  }

  // This post run loop is for logging of what happened and some data clean up
  for (let e in summary.run.executions) {
    logMessage(
      ` - Completed request '${summary.run.executions[e].item.name}' in ${summary.run.executions[e].response.responseTime} ms`
    )

    // Junk we don't want in data
    summary.run.executions[e].response.stream = '*REMOVED*'

    for (let a in summary.run.executions[e].assertions) {
      if (summary.run.executions[e].assertions[a].error) {
        logMessage(
          `ERROR! Request '${summary.run.executions[e].item.name}' - assertion failed: ${summary.run.executions[e].assertions[a].error.test}, Reason: ${summary.run.executions[e].assertions[a].error.message}`
        )

        // Junk we don't want in data
        summary.run.executions[e].assertions[a].error.message = '*REMOVED*'
        summary.run.executions[e].assertions[a].error.stack = '*REMOVED*'
      }
    }
  }
  fs.writeFileSync('debug.tmp.json', JSON.stringify(summary, null, 2))

  const time = summary.run.timings.completed - summary.run.timings.started
  logMessage(`Run complete, and took ${time}ms`)

  runCount++
  iterationCount += summary.run.stats.iterations.total
  reqCount += summary.run.stats.requests.total

  if (err) {
    logMessage(`ERROR! Failed to run collection ${err}`)
  }
  resultSummary = summary
  collectionName = summary.collection.name
}

function addMetric(metrics, name, value, type = 'gauge', labels = []) {
  metrics += `# TYPE postman_${name} ${type}\n`

  let labelsClone = [...labels]
  labelsClone.push({ collection: collectionName })

  let labelStr = ''
  for (let label of labelsClone) {
    let key = Object.keys(label)[0]
    let value = Object.values(label)[0]
    labelStr += `${key}="${value}",`
  }
  labelStr = labelStr.replace(/,\s*$/, '')

  metrics += `postman_${name}{${labelStr}} ${value}\n\n`
  return metrics
}

function logMessage(msg) {
  console.log(`### ${new Date().toISOString().replace('T', ' ').substr(0, 16)} ${msg}`)
}
